### PR TITLE
xIELU activation function

### PR DESCRIPTION
**Pull request by AllenHaoHuang**
https://wandb.ai/saesara/apertus_300m_baseline
https://wandb.ai/saesara/apertus_1b_baseline

Usage:
In config, replace --swiglu with --xielu
Multiple ff hidden dim by 1.5 if changing from gated mlp to standard mlp

### CODE REVIEW

**Comment by boeschf** [comment]
Do you still want to include the cuda implementation https://github.com/nickjbrowning/XIELU/tree/vectorized ?

**Comment by AllenHaoHuang** [comment]
It would be better to include it over the current implementation, but I thought it did not support BF16 yet

File: megatron/core/optimizer/__init__.py [hunk 1]
```diff
@@ -93,6 +93,10 @@ def _get_param_groups(
                 # Do not regularize biases and norm parameters.
                 no_wd = name.endswith(".bias") or len(param.shape) == 1
 
+            # TODO add more elegant way to disable weight decay for alpha_p and alpha_n
+            if "alpha_p" in name or "alpha_n" in name or "power" in name:
```
Comment (TJ-Solergibert) [inline megatron/core/optimizer/__init__.py:97]:
```suggestion
           if any(keyword in name for keyword in ["alpha_p", "alpha_n", "power"]):
```

File: megatron/training/activations.py [hunk 2]
```diff
@@ -1,8 +1,184 @@
...
+import torch.nn as nn
 import torch.nn.functional as F
 
+from megatron.core import parallel_state
+from megatron.core.dist_checkpointing import ShardedTensor
 from megatron.core.jit import jit_fuser
+from megatron.core.transformer.module import MegatronModule
+
+
+# Trying to apply @jit_fuser / @torch.compile to XIELU class causes issues with sharded_state_dict naming
```
Comment (TJ-Solergibert) [inline megatron/training/activations.py:12]: If it causes issues why it's ON?

File: megatron/training/arguments.py [hunk 3]
```diff
@@ -871,6 +871,12 @@ def core_transformer_config_from_args(args, config_class=None):
     if args.squared_relu:
         assert not args.swiglu
         kw_args['activation_func'] = squared_relu
+    if args.xielu:
+        kw_args['activation_func'] = XIELU
+    if args.xiprelu:
+        kw_args['activation_func'] = XIPReLU
+    if args.xiprelup:
+        kw_args['activation_func'] = XIPReLUP
```
Comment (TJ-Solergibert) [inline megatron/training/arguments.py:879]: Add safe guards to only use 1. What happens if I set `--swiglu --xielu --xiprelup` ?

Comment (AllenHaoHuang) [reply megatron/training/arguments.py:882]: added safeguards

File: megatron/training/activations.py (hunk 2 above)
Comment (AllenHaoHuang) [reply megatron/training/activations.py:12]: @torch.compile modifies parameter names (adds something like _orig_mod to the prefix). This caused issues with saving and loading checkpoints when using sharded state dict.

So I am only applying @torch.compile to the forward function rather than the whole XIELU class as a workaround.

File: megatron/training/arguments.py (hunk 3 above)
Comment (TJ-Solergibert) [reply megatron/training/arguments.py:882]: I see, but still, what happens with `kw_args['bias_activation_fusion']`?

Comment (AllenHaoHuang) [reply megatron/training/arguments.py:882]: Not implemented for xIELU or the Squared ReLU originally in the codebase

For SwiGLU/GELU its in megatron/core/fusions

I originally thought it was for CUDA fused implementations which isn't ready yet for xIELU. But for the activations, it looks like it is just fusing the activation with the bias from fc1. I can implement it if necessary, but I don't think it will be used?

Comment (TJ-Solergibert) [reply megatron/training/arguments.py:882]: Yes, but, shouldn't we set `kw_args['bias_activation_fusion']` to false? As it is true by default for both SwiGLU/GELU. What do you think?

**Review by TJ-Solergibert — APPROVED** [review]
👌
